runtime/dart_isolate.cc

```diff
@@ -803,10 +799,22 @@ DartIsolate::CreateDartVMAndEmbedderObjectPair(
 // |Dart_IsolateShutdownCallback|
 void DartIsolate::DartIsolateShutdownCallback(
     std::shared_ptr<DartIsolate>* embedder_isolate) {
+  // Terminate child Isolates
+  if ((*embedder_isolate)->terminate_child_isolates_) {
+    for (std::weak_ptr<DartIsolate> child_isolate :
+         (*embedder_isolate)->child_isolates_) {
+      if (!child_isolate.expired() && child_isolate.lock() != nullptr) {
+        bool shutdown_success = child_isolate.lock()->Shutdown();
```

Comment:
This code introduces a subtle timing error:

The issues:
* The expired check as I already said is redundant as locking a non-expired weak pointer gives you a shared pointer which extends the lifetime of the object in question for the duration of lifetime of the shared pointer.
* You are actually making two lock calls. In both cases, you are going to get two shared pointers that may both be invalid. In the first case, you end up with the unnamed temporary that get destroyed at the end of the conditional expression. But you checked it, so that fine. But the second lock call has no check! So if the object got destroyed after the first lock but before the second, you will end up dereferencing  null when you try to shutdown the isolate.

So instead of doing three checks with a timing issue, I suggest just one like so:

```
if (auto isolate = child_isolate.lock()) {
  isolate->Shutdown();
}
```